src/local/butler/reproduce.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""reproduce.py reproduces test cases locally."""
+
+from __future__ import print_function
+from builtins import object
+
+import os
+import shutil
+import tempfile
+
+from src.python.bot.tasks import commands
+from src.python.fuzzing import tests
+from src.python.system import environment
+from src.python.system import shell
+
+
+class _SimplifiedTestcase(object):
+  """Minimal representation of a test case."""
+
+  def __init__(self, testcase_json):
+    self.crash_state = testcase_json['crash_state']
+    self.security_flag = testcase_json['security_flag']
+    self.gestures = testcase_json['gestures']
+    self.flaky_stack = testcase_json['flaky_stack']
+
+    # Custom field not included in real test cases. Used in environment setup.
+    self.job_definition = testcase_json['job_definition']
+
+
+def _get_testcase(_):
+  """Retrieve the json representation of the test case with the given id."""
+  # TODO(mbarbella): Actually fetch the test case info from the server.
+  testcase_json = {
+      'crash_state': '',
+      'security_flag': False,
+      'gestures': [],
+      'flaky_stack': False,
+      'job_definition': 'APP_NAME = echo\nAPP_ARGS = -n\n',
+  }
+
+  return _SimplifiedTestcase(testcase_json)
+
+
+def _download_testcase(_):
+  """Download the test case and return its path."""
+  # TODO(mbarbella): Implement this.
+  return '/tmp/blah'
+
+
+def _prepare_environment(testcase, build_directory):
+  """Prepare environment variables based on the test case and build path."""
+  # Create a temporary directory to use as ROOT_DIR with a copy of the default
+  # bot and configuration directories nested under it.
+  old_root_dir = environment.get_value('ROOT_DIR')
+  root_dir = tempfile.mkdtemp()
+  shutil.copytree(
```

Comment:
I don't think this is enough, there is resources dir needed as well. I think we can remove the mkdtemp logic for now. Right now it is in gitignore, so those dirs wont conflict. When we later package reproduce tool, it won't matter either since that dir will be temp in itself. then can remove line 96 as well.

Reply:
It seems to delete the READMEs without it, which is inconvenient. I'll see if I can resolve this now.

Reply:
Added the resources directory.